packages/devtools_app/lib/src/config_specific/drag_and_drop/drag_and_drop.dart

```diff
@@ -2,61 +2,169 @@
+  /// the (x, y) coordinates, and updates the active state for all [DragAndDrop]
+  /// widgets accordingly.
+  void hitTestAndUpdateActiveId(double x, double y) {
+    final hitTestResult = HitTestResult();
+    RendererBinding.instance.hitTest(hitTestResult, Offset(x, y));
+
+    // Starting at bottom of [hitTestResult.path], look for [DragAndDrop] or
+    // [DragAndDropEventAbsorber] widgets. These widgets will be marked by a
+    // [RenderMetaData] target with [DragAndDropMetaData] metaData.
+    for (final result in hitTestResult.path) {
+      if (result.target is RenderMetaData) {
+        final target = result.target as RenderMetaData;
+        // The first [DragAndDropMetaData] we find will either be for the active
+        // [DragAndDrop] widget or for a [DragAndDropEventAbsorber] widget. We
+        // should return if we find either.
+        if (target.metaData is DragAndDropMetaData) {
+          final metaData = target.metaData as DragAndDropMetaData;
+
+          // Deactivate all drag and drop states if we find
+          // [_dragAndDropEventAbsorberId].
+          if (metaData.value == _dragAndDropEventAbsorberId) {
+            for (final state in _statesById.values) {
+              state.activate(false);
+            }
+            return;
+          }
+
+          // Otherwise, activate the first drag and drop state we found and
+          // deactivate all others.
+          final activeDragAndDropState = _statesById[target.metaData.value];
+          if (activeDragAndDropState != null) {
+            for (final state in _statesById.values) {
+              activeState = activeDragAndDropState;
+              state.activate(state == activeDragAndDropState);
+            }
+            return;
+          }
+        }
+      }
+    }
+  }
+}
+
+class DragAndDrop extends StatefulWidget {
+  const DragAndDrop({
+    @required this.id,
+    @required this.manager,
+    @required this.handleDrop,
+    @required this.child,
+  });
+
+  final String id;
 
   /// Callback to handle parsed data from drag and drop.
   ///
   /// The current implementation expects data in json format.
-  final void Function(Map<String, dynamic> data) handleDrop;
+  final DevToolsJsonFileHandler handleDrop;
+
+  final DragAndDropManager manager;
 
   final Widget child;
+
+  @override
+  State<DragAndDrop> createState() => DragAndDropState();
 }
 
-abstract class DragAndDropState extends State<DragAndDrop> {
+class DragAndDropState extends State<DragAndDrop> {
   final _dragging = ValueNotifier<bool>(false);
 
-  NotificationsState notifications;
+  bool _isActive;
+
+  @override
+  void initState() {
+    super.initState();
+    widget.manager.registerDragAndDrop(widget.id, this);
+  }
 
   @override
-  void didChangeDependencies() {
-    notifications = Notifications.of(context);
-    super.didChangeDependencies();
+  void dispose() {
+    super.dispose();
+    widget.manager.unregisterDragAndDrop(widget.id);
```

Comment:
by convention, call super.disppose() last.
@pq this would be a nice flutter specific lint.

Reply:
done